libs/feedback/src/lib/feedback-widget/feedback-widget.component.ts

```diff
@@ -1,50 +1,73 @@
-import { ActivatedRoute } from '@angular/router';
-import { Component, HostListener, OnInit } from '@angular/core';
+import { ActivatedRoute, Router } from '@angular/router';
+import { Component, OnInit } from '@angular/core';
 import { FeedbackService } from '../feedback.service';
 import { Message } from '../message';
 import { Observable } from 'rxjs';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { debounceTime } from 'rxjs/operators';
 
-function findMatchingDOMAncestor(element) {
-  while (element.parentNode) {
-    if (
-      element.className &&
-      element.className.indexOf('feedback-container') >= 0
-    ) {
-      return true;
-    }
-    element = element.parentNode;
-  }
-}
 
 @Component({
   selector: 'feedback-widget',
   templateUrl: './feedback-widget.component.html',
   styleUrls: ['./feedback-widget.component.css']
 })
 export class FeedbackWidgetComponent implements OnInit {
+
   messages$: Observable<Message[]>;
-  open: boolean;
+
+  formGroup: FormGroup;
+  statusMessage = '';
+  error = false;
 
   constructor(
+    private fb: FormBuilder,
+    private activatedRoute: ActivatedRoute,
     private feedbackService: FeedbackService,
-    private activatedRoute: ActivatedRoute
-  ) {}
+    private router: Router
+  ) {
+    this.router.events.subscribe(() => {
+      this.messages$ = this.feedbackService.getMessages(this.activatedRoute);
+    });
+  }
 
   ngOnInit() {
-    this.messages$ = this.feedbackService.getMessages(this.activatedRoute);
+    let value = localStorage[`feedback-${this.router.url}-comment`] || '';
+    value = value === 'null' ? '' : value;
+    this.formGroup = this.fb.group({
+      comment: [value, Validators.required],
+      name: [localStorage.getItem('userName') || '', Validators.required],
+      email: [localStorage.getItem('userEmail') || '', []]
+    });
+
+    this.formGroup.valueChanges.pipe(debounceTime(500)).subscribe(data => {
```

Comment:
missing unsubscribe to dispose subscription

Reply:
Like above about subscriptions